src/flatten.jl

```diff
@@ -0,0 +1,163 @@
+           val2 = (b1 = 11, b2 = 22),
+           val3 = [111, 222],
+           val4 = 4.3,
+       );
+
+julia> flatten(data)
+(val1⁻a1 = 1, val1⁻a2 = 2, val2⁻b1 = 11, val2⁻b2 = 22, val3 = [111, 222], val4 = 4.3)
+```
+
+    flatten(A, dims, [delim])
+
+Flatten a `KeyedArray` along the specified consecutive dimensions.
+The `dims` argument can either be a `Tuple` of symbols or a `Pair{Tuple, Symbol}` if
+you'd like to specify the desired flattened dimension name.
+If the `dims` is just a `Tuple` with no output dimension specified then $DEFAULT_PROD_DELIM
+will be used to generate the new dimension name.
+
+# Example
+```jldoctest
+julia> using AxisKeys, Dates, NamedDims; using AxisSets: flatten
+
+julia> A = KeyedArray(
+           reshape(1:24, (4, 3, 2));
+           time=DateTime(2021, 1, 1, 11):Hour(1):DateTime(2021, 1, 1, 14),
+           obj=[:a, :b, :c],
+           loc=[1, 2],
+       );
+
+julia> dimnames(flatten(A, (:obj, :loc)))
+(:time, :objᵡloc)
+```
+"""
+function flatten end
+
+# Single arg version defaults to `nothing` delim
+# All other methods dispatch on the two arg form to be explicit
+flatten(x) = flatten(x, nothing)
+
+# Primary flatten algorithm. Other methods should call this and adjust keys as necessary.
+function flatten(x::Union{Vector{<:Pair}, Iterators.Pairs}, delim::Nothing)
+    result = Channel{Pair}() do chnl
+        for (k, v) in x
+            if isassociative(v)
+                children = flatten(v)
+                iter = isa(children, NamedTuple) ? pairs(children) : children
+                for (_k, _v) in iter
+                    new_key = isa(_k, Tuple) ? (k, _k...) : (k, _k)
+                    put!(chnl, new_key => _v)
+                end
+            else
+                put!(chnl, (k,) => v)
+            end
+        end
+    end
+
+    return collect(result)
+end
+
+# NOTE: NamedTuples only support symbol names, so we use a simple :⁻ delimiter
+function flatten(x::NamedTuple, delim::Symbol=DEFAULT_FLATTEN_DELIM)::NamedTuple
+    kwargs = map(flatten(pairs(x))) do (k, v)
+        _k = isa(k, Tuple) ? join(k, delim) : k
+        return Symbol(_k) => v
+    end
+    return (; kwargs...)
+end
+
+# Since the key type could be changed as part of flattening we can only ensure the that
+# result is a `Dict` (vs the same input dict).
+flatten(x::AbstractDict, delim) = Dict(flatten(collect(x), delim)...)
+
+# Utility flatten to handle joining tuples of symbols or strings into a single string/symbol
+function flatten(x::Vector{<:Pair{T}}, delim::T) where T<:AbstractString
+    return [join(k, delim) => v for (k, v) in flatten(x)]
+end
+
+function flatten(x::Vector{<:Pair{Symbol}}, delim::Symbol)
+    return [Symbol(join(k, delim)) => v for (k, v) in flatten(x)]
+end
+
+function flatten(A::XArray, dims::Tuple, delim=DEFAULT_PROD_DELIM)
+    new_name = Symbol(join(dims, delim))
+    flatten(A, dims => new_name, delim)
+end
+
+function flatten(A::XArray, dims::Pair{<:Tuple, Symbol}, delim=nothing)
```

Comment:
is the fact we allow/default to nothing to enable the new keys be represented as tuples `(:a, 1)` etc?

Reply:
Yes, the goal is to use `Tuple` keys most of the time. If we're given an explicit string/symbol delimiter, or we're operating on `NamedTuple`s (require symbols) then we handle those separately. In the case of `NamedTuple`s, I figured it's better to return the same type back rather than forcing the `Tuple` key default.